File: Python-2.7/CodeNamesCard.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
"""
Code Names Card
@author: ivory
"""

import argparse
import random
import numpy as np
from matplotlib.patches import Rectangle
from matplotlib.collections import PatchCollection
import matplotlib.pyplot as plt
from PIL import Image
import sys


def retrieve_args():

	parser = argparse.ArgumentParser(description='Create a spy map card like the one used for the Code Names board game.')
	# height
	parser.add_argument("-H", "--height", metavar='H', type=int, default=None, nargs='?',
	                    help='number of rows in the grid (default 5 or equal to width)')
	# width
	parser.add_argument('-W', "--width", metavar='W', type=int, default=None, nargs='?',
	                    help='number of columns in the grid (default 5 or equal to height)')
	# assassin
	parser.add_argument('-a', "--assassin", metavar='a', type=int, default=1, nargs='?',
	                    help='number of assissins')
	# inocent bystanders
	parser.add_argument('-i', "--incents", metavar='i', type=int, default=None, nargs='?',
	                    help='number of innocent by-stander squares', dest="ib")
	# blue team
	parser.add_argument('-b', "--blue", metavar='b', type=int, default=None, nargs='?',
	                    help='number of blue team squares')
	# red team
	parser.add_argument('-r', "--red", metavar='r', type=int, default=None, nargs='?',
	                    help='number of red team squares')
	# output file name
	parser.add_argument('-o', "--outfile", metavar='i', type=str, nargs='?',
	                    help='file name to save image (default: SpyMap.png)', 
	                    default="SpyMap.png",)
	# random seed
	parser.add_argument('-s', "--set-seed", metavar='s', type=int, default=None, nargs='?',
	                    help='set random seed to regenerate an identical card', dest="ss")

	args = parser.parse_args()

	return args


def process_grid_size(args):
	# if both height and width are null, set both to 5.
	# if only one is NULL, set it to match the other.
	if args.height is None and args.width is None:
		args.height = 5
		args.width = 5
	elif args.height is None:
		args.height = args.width
	elif args.width is None:
		args.width = args.height

	ts = args.width * args.height
	return args, ts



def process_square_types(args, ts):
	# Notice that the order of handling cases matters
	# first - if both blue and red are specified
	# second - if neighter are specified 
	#     (if they were specified, this is correctly skipped)
	# last - if either one is null 
	#     (which must be exactly one, since both values would have 
	#     been set by now if both or neither had been specified.)
	#     so we can assume that the other is NOT None
	first = ""

	# if neither -b nor -r is None (both are specified)
	if args.blue is not None and args.red is not None:
		# calculate open squares, total - assassin - red - blue
		openSq = ts - args.assassin - args.blue - args.red
		if openSq < 0:
			sys.exit("Not enough squares in the grid.\n")
		if args.ib is not None and args.ib != openSq:
			sys.exit("This doesn't add up. Check the number of squares you need.")
		if args.ib is None:
			args.ib = openSq
		if args.blue >= args.red:
			first = "blue"
		else: 
			first = "red"

	# if both -b and -r are None
	if args.blue is None and args.red is None:
		#if -i is None, calculate -i as 25% of total squares
		if args.ib is None:
			args.ib = ts // 4
		# Calculate open squares
		openSq = ts - args.assassin - args.ib
		## Make sure open squares is odd
		# One team has to go first, the other team has to have 
		# a -1 addvantage to make up for it.
		# If the number of open squares is even, 
		# make it odd by adding or subtracting one innocent
		if openSq < 3:
			sys.exit("Not enough squares in the grid.\n")
		if openSq % 2 == 0:
			if args.ib > openSq / 2:
				args.ib = args.ib - 1
				openSq = openSq +1
			else:
				args.ib = args.ib + 1
				openSq = openSq - 1
		# randomely pick red or blue to be 'first' the other 'second'
		sample = random.sample(["blue", "red"], 2)
		first = sample[0]
		second = sample[1]
		# assign first to get open squares /2 rounded up, and second openSq rounded down
		if first is "blue":
			args.blue = (openSq // 2) + 1
			args.red = openSq // 2
		else:
			args.red = (openSq // 2) + 1
			args.blue = openSq // 2

	# if exactly one of red or blue is none,
	# assign the one that is specified to be first, the other second
	# assign second to get first - 1 squares
	if args.blue is None or args.red is None:
		if args.red is not None:
			first = "red"
			args.blue = args.red - 1
		elif args.blue is not None:
			first = "blue"
			args.red = args.blue - 1
		# Calculate open squares
		if args.blue is not None and args.red is not None:
			openSq = ts - args.assassin - args.red - args.blue
		# if <0, error meesage
		if openSq < 0:
			sys.exit("Not enough squares in the grid.\n")
		if openSq == 0:
			print "There are no inocent by-standers on this map.\t" #TODO convert this to a proper warning
		if args.ib is None:
			args.ib = openSq
		if args.ib is not None and args.ib != openSq:
			sys.exit("This doesn't add up.")
	return args, first




def main():

	args = retrieve_args()

	if args.ss is not None:
		random.seed(args.ss)
		# later use random.sample(population, k) to sample without replacement

	args, ts = process_grid_size(args)
	args, first = process_square_types(args, ts)

	gap = .1
	border = .3

	slotIds = np.array(random.sample(range(ts), ts))
	columns = slotIds % args.width + border
	rows = slotIds // args.width + border

	square_type = ["assassin"] * args.assassin + ["ib"] * args.ib + ["blue"] * args.blue + ["red"] * args.red
	types_dict = {slotIds[i]:square_type[i] for i in range(ts)}

	colors = ["black"] * args.assassin + ["tan"] * args.ib + ["blue"] * args.blue + ["red"] * args.red
	colors_dict = {slotIds[i]:colors[i] for i in range(ts)}

	plot_width = border+args.width-gap+border
	plot_height = border+args.height-gap+border

	patches = []
	for i in range(ts):
		space = Rectangle([columns[i], rows[i]], width=(1 - gap), height=(1 - gap)) # colors_dict.values()
		patches.append(space)
	pc = PatchCollection(patches, color=colors_dict.values())

	#### Define a line to border the image, and colored rectangles 
	# to indicate which team goes first.
	# Notice that the size of this box is equal to the diminsions of 
	# the whole plot, minus half the border on each side.
	outer_patches = []
	outline = Rectangle([border / 2.0, border / 2.0], width=args.width-gap+border, height=args.height-gap+border, fc='none', ec='#000000')
	outer_patches.append(outline)
	shortSide = .2
	longSide = .6
	outer_widths = [shortSide, longSide, shortSide, longSide] #left, top, right, bottom
	outer_heights = [longSide, shortSide, longSide, shortSide]
	# set the x values for the lower LEFT corner for the rectangles on the border
	outer_x = [(border / 2.0 - shortSide / 2.0), + #left
	(plot_width - longSide)/2.0, + #top
	plot_width - border / 2.0 - shortSide / 2.0, + # right
	(plot_width - longSide)/2.0] # bottom
	# set the y values for the LOWER left corner for the rectangles on the border
	outer_y = [(plot_height - longSide)/2.0, + #left
	plot_height - border / 2.0 - shortSide / 2.0, + #top
	(plot_height - longSide)/2.0, + # right
	border / 2.0 - shortSide / 2.0] # bottom

	for i in range(4):
		space = Rectangle([outer_x[i], outer_y[i]], width=outer_widths[i], height=outer_heights[i], fc=first, ec='#000000')
		outer_patches.append(space)
	outer_pc = PatchCollection(outer_patches, match_original=True)

	#plt.ion() #for dev and testing
	fig, ax = plt.subplots()

	# I the plot itself to make up the entire image, no surrounding white space.
	fig.subplots_adjust(left=0, bottom=0, right=1, top=1)
	
	ax.add_collection(pc)
	ax.add_collection(outer_pc)
	ax.set_xlim(left=0, right=plot_width)
	ax.set_ylim(bottom=0, top=plot_height)

	plt.savefig(args.outfile)
	img = Image.open(args.outfile)
	img.show()
	#plt.show()


	print "total squares in grid: %d" % ts
	print args
	print "%s team to go first." % first




if __name__ == "__main__": main()

File: Java/SpyCard/src/com/company/CodeNamesCard.java
package com.company;

import org.apache.commons.cli.CommandLine;
import org.apache.commons.cli.Option;

import java.util.Random;

import java.util.HashMap;
import java.lang.Math;

public class CodeNamesCard {

    // in case we need one
//    public Random rand = new Random();
    public HashMap<String, Integer> opt;
    public HashMap<String, String> stringOptions;
    private int height;
    private int width;
    private int totalSquares;
    private int assassin;
    private int inocents;
    private int red;
    private int blue;
    private String first;
    private String second;


    public int getHeight() {
        return height;
    }

    public int getWidth() {
        return width;
    }

    public int getTotalSquares() {
        return totalSquares;
    }

    public int getAssassin() {
        return assassin;
    }

    public int getInocents() {
        return inocents;
    }

    public int getRed() {
        return red;
    }

    public int getBlue() {
        return blue;
    }

    public String getFirst() {
        return first;
    }

    public String getSecond() {
        return second;
    }

    CodeNamesCard(CommandLine cmd) throws Exception {
        Random rand = new Random();
        String[] rb = {"r", "b"};

        /*TODO: instead of making these HashMaps,
        just take the options directly from the commandLine object and set the fields.
        Then refactor to query the fields, not the HashMaps,
        And remove the bit at the bottom that transfers the values from the HashMaps to the fields.
         */
        HashMap<String, Integer> opt = new HashMap<String, Integer>();
        HashMap<String, String> stringOpt = new HashMap<String, String>();

        Option[] processedOpts = cmd.getOptions();
        for (Option po : processedOpts) {
//            System.out.println(po.getOpt());
//            System.out.println(po.getValue());
            opt.put(po.getOpt(), Integer.parseInt(po.getValue()));
        }

        // defaults - if it doesn't have a value, give it this value
        opt.putIfAbsent("a", 1);
        //stringOpt.putIfAbsent("o", "SpyMap.png");


        //If -s is not null (and not an empty string), set the seed.
        if (opt.containsKey("s")) {
            rand.setSeed(opt.get("s"));
        }
        //TODO: if s is not specified, pick a random int and use that---so any card can be recreated.
        //TODO: move the set seed stuff to main to be passed to CardLayout

        // if both height and width are absent, set both to 5.
        // if only one is absent, set it to match the other.
        // if both are specified, great!--do nothing.
        if (!opt.containsKey("H") && !opt.containsKey("W")) {
            opt.put("H", 5);
            opt.put("W", 5);
        } else {
            if (!opt.containsKey("H")) {
                opt.put("H", opt.get("W"));
            }
            if (!opt.containsKey("W")) {
                opt.put("W", opt.get("H"));
            }
        }

        // Calculate total squares
        opt.put("ts", (opt.get("H") * opt.get("W")));

        // Set the number of red and blue squares
        /*
        Notice that the order of handling cases matters
	    first - if both blue and red are specified
        second - if neighter are specified
	        (if they were specified, this is correctly skipped)
	    last - if either one is absent
	        (which must be exactly one, since both values would have
	        been set by now if both or neither had been specified.)
	        so we can assume that the other is NOT absent
         */

        // if neither -r nor -b is absent // ie, if both are specified
        if (opt.containsKey("b") && opt.containsKey("r")) {
            // Here, we assume that both b and r are specified.
            // the team with more squares goes first
            stringOpt.putIfAbsent("first", (opt.get("b") > opt.get("r")) ? "b" : "r");
            // if both teams have the same number, choose randomly
            //TODO: test this
            if (opt.get("b").equals(opt.get("r"))) {
                //String[] rb = {"r", "b"};
                stringOpt.putIfAbsent("first", rb[rand.nextInt(1)]);
            }
            // calculate open squares, total - assassin - red - blue
            // this should never be less than 0.
            opt.put("openSq", opt.get("ts") - opt.get("a") - opt.get("r") - opt.get("b"));
            if (opt.get("openSq") < 0) {
                throw new java.lang.Error("Not enough squares in the grid.");
            }
            // if i is also specified AND its equal to open squares --> great!
            // if i is also specified AND its not equal to open squares, then the user has tied a knot.
            if (opt.containsKey("i") && !opt.get("i").equals(opt.get("openSq"))) {
                throw new java.lang.Error("This doesn't add up.");
            }
            // if i is not specified, set it equal to open squares
            if (!opt.containsKey("i")) {
                opt.put("i", opt.get("openSq"));
            }
            //end the assumption that both b and r are specified.
        }

        // if both are absent
        if (!opt.containsKey("b") && !opt.containsKey("r")) {
            // we assume that both are absent
            // if -i is also absent, calculate i as 25% of total squares
            opt.putIfAbsent("i", (int) Math.floor(opt.get("ts") * .25));
            // calculate open squares as total - assassin - i
            opt.put("openSq", opt.get("ts") - opt.get("a") - opt.get("i"));
            // if open squares is less than 3, then you don't have room for one blue, red and assassin.
            if (opt.get("openSq") < 3) {
                throw new java.lang.Error("Not enough squares in the grid.");
            }
            /* Make sure open squares is odd
		    One team has to go first, the other team has to have
		    a -1 addvantage to make up for it.
		    If the number of open squares is even,
		    make it odd by adding or subtracting one innocent
		    */
            if (opt.get("openSq") % 2 == 0) { //if open squares is even
                if (opt.get("i") > opt.get("openSq") / 2) {
                    //take one from innocent by-standars and give it to open squares
                    opt.put("i", opt.get("i") - 1);
                    opt.put("openSq", opt.get("i") + 1);
                } else {
                    opt.put("i", opt.get("i") + 1);
                    opt.put("openSq", opt.get("openSq") - 1);
                }
            }
            // randomly pick red or blue to be "first", the other second
            int rint = rand.nextInt(2);
            stringOpt.put("first", rb[rint]);
            // make the other one second
            stringOpt.put("second", stringOpt.get("first").equals("r") ? "b" : "r");
            // assign 'first' to get open squares/2 rounded up
            opt.put(stringOpt.get("first"), (int) Math.floor(opt.get("openSq") / 2)+1 ); //not sure why ceil didn't work
            // assign 'second' to get open squares/2 rounded down
            opt.put(stringOpt.get("second"), (int) Math.floor(opt.get("openSq") / 2));
            // end the assumption that both red and blue are absent
        }

        // if exactly one is absent
        if (!opt.containsKey("r") | !opt.containsKey("b")){
            // Here we assume that one but not the other was specified
            // assign the one that is not null to be first, the other second
            // assign the second one to get first - 1 squares
            if (!opt.containsKey("b")){
                stringOpt.put("first", "r");
                stringOpt.put("second", "b");
                opt.put("b", opt.get("r") - 1);
            }else{ //ie, if blue is specified and red is not
                stringOpt.put("first", "b");
                stringOpt.put("second", "r");
                opt.put("r", opt.get("b") - 1);
            }
            // calculate open squares; this should not be less than 0
            opt.put("openSq", opt.get("ts") - opt.get("a") - opt.get("r") - opt.get("b"));
            if (opt.get("openSq") < 0) {
                throw new java.lang.Error("Not enough squares in the grid.");
            }
            if (opt.get("openSq") == 0){
                throw new java.lang.Exception("There are no inocent by-standers on this map.");
            }
            // if i is absent, set it equal to open squares
            opt.putIfAbsent("i", opt.get("openSq"));
            // if i is specified and it is not equal to open squares, that's bad
            if (opt.containsKey("i") && !opt.get("i").equals(opt.get("openSq"))) {
                throw new java.lang.Error("This doesn't add up.");
            }
        // end assumption that one but not the other was specified
        }

//        this.opt = opt;
//        this.stringOptions = stringOpt;
        this.height = opt.get("H");
        this.width = opt.get("W");
        this.totalSquares = opt.get("ts");
        this.assassin = opt.get("a");
        this.inocents = opt.get("i");
        this.red = opt.get("r");
        this.blue = opt.get("b");
        this.first = stringOpt.get("first");
        this.second = stringOpt.get("second");

    }

    @Override
    public String toString() {
        return "CodeNamesCard{" +
                "height=" + height +
                ", width=" + width +
                ", total squares=" + totalSquares +
                ", assassin=" + assassin +
                ", inocents=" + inocents +
                ", red=" + red +
                ", blue=" + blue +
                ", first='" + first + '\'' +
                ", second='" + second + '\'' +
                '}';
    }
}
